src/api/mc_worker_api.erl

```diff
@@ -274,3 +486,308 @@ assign_id(Doc) ->
     {} -> bson:update(<<"_id">>, mongo_id_server:object_id(), Doc);
     _Value -> Doc
   end.
+
+%% Source https://www.mongodb.com/docs/manual/reference/command/
+%% Might need to get updated if commands are added
+get_existing_commands_map() ->
```

Comment:
Thanks for the suggestion. Are you sure this leads to better performance? I think this will give a time complexity of O(number_of_commands). For example, when there is no match for the argument the Erlang emulator will look at all the function clauses (I think). A look up in a map will give a time complexity of O(log(number_of_commands)) which I think should be much better since there are quite a lot of commands.

Reply:
Erlang function clause matches are tree-structured, the worst case it's log2 complexity.

Reply:
> Erlang function clause matches are tree-structured, the worst case it's log2 complexity.

Thanks, interesting, I did not know this.

I see what you mean now. If the map is rebuilt every time, it is not great code. I though that since the map is a literal (it does not contain any dynamically computed data) it would be put in the literal storage only once when the module is loaded so that it would only need to be built once. My bad for just assuming this and not checking if this is actually the case.

I wrote a small benchmark to compare the two approaches. See:

https://gist.github.com/kjellwinblad/dbab63d3aa1bee91ade1622f66f3e1c6

When I run this it seems like using the approach you suggest is about twice as fast as using my literal map approach. I think this must mean that the map is built only once when the module is loaded since I think it would be much slower if it had to be rebuilt every time one calls the get_existing_commands_map function.

Anyway, I will change the code to use the function clause approach as this might be the conventional way of doing this and is a little bit faster.

Reply:
Which is more commonly used? atom or binary?
maybe we only need to implement the more commonly used type and convert the other type to check.

if we have to implement both, we can split the implementation one for `when is_atom()` and one for `when is_binary()`, and add a eunit to make sure they both have the exact same enumerations.

Reply:
Just to make sure, I added a third alternative to the benchmark that puts the map in persistent_term storage. This alternative performs about 10% worse than the literal map acquired with the function. I think this is expected since there is some overhead of looking up the persistent_term key in the persistent_term storage. This confirms that literal maps are just constructed once.

Reply:
> Which is more commonly used? atom or binary?

This is a public API so it is difficult to say.

> maybe we only need to implement the more commonly used type and convert the other type to check.

I think this is a good idea so we don't have "duplicates" in the "list". Since it is a public API, I don't think we want to convert binaries to atoms (the atom table has a fixed size) so maybe the other way around is better.

> if we have to implement both, we can split the implementation one for when is_atom() and one for when is_binary(), and add a eunit to make sure they both have the exact same enumerations.

Not sure about splitting the implementation as I'm unsure what benefit that would give but an eunit test to check that there is always a pair would be good but I think your first suggestion is even better even though it might have a small performance price. That being said, I think the whole library is quite poorly tested so I think there are more low hanging fruits that should be prioritized before.